crates/polars-ops/src/frame/join/asof/default.rs

```diff
@@ -259,10 +258,15 @@ pub(super) fn join_asof_nearest_with_tolerance<
                     break;
                 }
             } else {
-                // We'ved moved farther away, so the last element was the match.
-                out.push(Some(offset - 1));
+                // We'ved moved farther away, so the last element was the match if it's within tolerance
```

Comment:
The code previously was simply accepting that the prior match was a success without making sure it was eligible, ditto for the groups version below.